tests/unittest.py

```diff
@@ -124,6 +129,74 @@ def new(*args: P.args, **kwargs: P.kwargs) -> R:
     return _around
 
 
+@overload
+def deepcopy_config(config: RootConfig, root: Literal[True]) -> RootConfig:
+    ...
+
+
+@overload
+def deepcopy_config(config: Config, root: Literal[False]) -> Config:
+    ...
+
+
+def deepcopy_config(
+    config: Union[Config, RootConfig], root: bool
+) -> Union[RootConfig, Config]:
+    new_config: Union[Config, RootConfig]
+
+    if root:
+        typed_rootconfig = cast(RootConfig, config)
+        new_config = typed_rootconfig.__class__(typed_rootconfig.config_files)
+    else:
+        typed_config = cast(Config, config)
+        new_config = typed_config.__class__(typed_config.root)
+
+    for attr_name in config.__dict__:
+        if attr_name.startswith("__") or attr_name == "root":
+            continue
+        attr = getattr(config, attr_name)
+        if isinstance(attr, Config):
+            new_attr = deepcopy_config(attr, root=False)
+        else:
+            new_attr = attr
+
+        setattr(new_config, attr_name, new_attr)
+
+    return new_config
+
+
+_make_homeserver_config_obj_cache: Dict[str, Union[RootConfig, Config]] = {}
+
+
+def make_homeserver_config_obj(config: Dict[str, Any]) -> RootConfig:
+    """Creates a :class:`HomeServerConfig` instance with the given configuration dict.
+
+    This is equivalent to::
+
+        config_obj = HomeServerConfig()
+        config_obj.parse_config_dict(config, "", "")
+
+    but it keeps a cache of `HomeServerConfig` instances and deepcopies them as needed,
+    to avoid validating the whole configuration every time.
```

Comment:
I don't follow why we're doing the deepcopying. Is it to allow tests to mutate config at test-time?

Reply:
Or, put differently: what sort of validation are we avoiding here?

Reply:
jsonschema and compiling Jinja templates. Profiling interfers heavily with the runtimes there, so I cannot easily get exact timings.

Reply:
I wonder if its easier to return the config wrapped in a class that stores any modifications on itself, e.g.:

```python
class WrappedConfig:
    def __init__(self, config):
        self.config = config

    def __getattr__(self, name):
        return getattr(self.config, name)
```

that way you avoid having to try and write a deep copy mechanism, while ensuring the underlying config object doesn't change?

Reply:
This would only work when setting attributes directly to the root attribute. Otherwise, `root.foo.bar = "abc"`  would add an attribute to the object `root.foo`, but we still need to somehow get rid of it when resetting `root`.

Reply:
Ah yup, you're totally right.